File: bin/memo.rb
#!/usr/bin/ruby

require 'optparse'

DEFAULT_DIR = "#{Dir.home}/doc/memo".freeze

def params
  hash = {}
  opt = OptParse.new
  opt.on('--config [PATH]') { |path| hash[:config] = path || true }
  opt.on('--grep KEYWORD') { |keyword| hash[:grep] = keyword }
  args = opt.parse(ARGV)
  { args: args, opts: hash }
end

def open_memo(path)
  exec("vim \"#{path}\" < /dev/tty")
end

def action(params)
  opts = params[:opts]
  work_dir = DEFAULT_DIR

  # script exits after echo current work directory path.
  exec("echo #{work_dir}") if opts[:config] == true

  work_dir = opts[:config] if opts[:config]

  # script exits after grep with option parameter.
  exec("ag #{opts[:grep]} #{work_dir}") if opts[:grep]

  filename = params[:args].first

  if filename.nil?
    if Dir.exist?(work_dir)
      # browse file with fzf
      select_filename = `ls #{work_dir} | fzf --preview "head -100 #{work_dir}/{}"`
      open_memo("#{work_dir}/#{select_filename.chomp}") unless select_filename.empty?
    end
  else
    exit unless system("mkdir -p #{work_dir}")
    open_memo("#{work_dir}/#{filename}.md")
  end
end

action(params)

File: bin/memo.py
#! /usr/bin/env python

import sys
import subprocess
import argparse
from pathlib import Path

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("name", nargs="?")
    parser.add_argument("--config", nargs="?", const=False)
    parser.add_argument("-g", "--grep")
    return parser.parse_args()

def default_memo_dir():
    return Path.home().joinpath("doc/memo")

def memo_dir(current_path):
    return Path(current_path)

def agrep(word, work_dir):
    subprocess.call(["ag", word, work_dir])

def make_dir(work_dir):
    subprocess.call(["mkdir", "-p", work_dir])

def open_memo(filename, work_dir):
    subprocess.call(["vim", filename], cwd=work_dir)

def select_file_with_fzf(work_dir):
    proc = subprocess.run(f"ls {work_dir} | fzf --preview \"head -100 {work_dir}/{{}}\"", shell=True, stdout=subprocess.PIPE)
    return proc.stdout.decode("utf8").rstrip('\n')

def main():
    args = parse_args()

    # config
    if args.config == False:
        print(default_memo_dir())
        sys.exit()

    work_dir = memo_dir(args.config) if args.config else default_memo_dir()

    # grep
    if args.grep:
        agrep(args.grep, work_dir)
        sys.exit()

    # new
    if args.name:
        make_dir(work_dir)
        open_memo(f"{args.name}.txt", work_dir)

    # edit
    elif work_dir.is_dir():
        filename = select_file_with_fzf(work_dir)
        if filename:
            open_memo(filename, work_dir)

if __name__ == '__main__':
    main()
